File: source/bindings/java/com/microsoft/cognitiveservices/speech/NoMatchDetails.java
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//
package com.microsoft.cognitiveservices.speech;

import com.microsoft.cognitiveservices.speech.util.Contracts;
import com.microsoft.cognitiveservices.speech.util.IntRef;
import com.microsoft.cognitiveservices.speech.util.SafeHandle;

/**
 * Contains detailed information for NoMatch recognition results.
 */
public class NoMatchDetails {

    private com.microsoft.cognitiveservices.speech.NoMatchReason reason;

    /**
     * Creates an instance of NoMatchDetails object for the NoMatch SpeechRecognitionResults.
     * @param result The recognition result that has NoMatch returned in Reason.
     * @return The NoMatchDetails object being created.
     */
    public static com.microsoft.cognitiveservices.speech.NoMatchDetails fromResult(RecognitionResult result) {
        return new NoMatchDetails(result);
    }

    /*! \cond PROTECTED */

    protected NoMatchDetails(RecognitionResult result) {
        Contracts.throwIfNull(result, "result");

        IntRef intVal = new IntRef(0);
        Contracts.throwIfFail(getResultReason(result.getImpl(), intVal));
        this.reason = NoMatchReason.values()[(int)intVal.getValue() - 1]; // Native NoMatchReason enum starts at 1!!
    }

    /*! \endcond */

    /**
     * Explicitly frees any external resource attached to the object
     */
    public void close() {
    }

    /**
     * The reason why NoMatch is returned.
     * @return Specifies the reason for NoMatch.
     */
    public com.microsoft.cognitiveservices.speech.NoMatchReason getReason() {
        return this.reason;
    }

    /**
     * Returns a String that represents the no match details.
     * @return A String that represents the no match details.
     */
    @Override
    public String toString() {
        return "NoMatchReason:" + this.reason;
    }

    private final native long getResultReason(SafeHandle resultHandle, IntRef reasonVal);
}

File: source/bindings/js/src/sdk/NoMatchDetails.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

import {
    RecognitionStatus,
    SimpleSpeechPhrase,
} from "../../src/common.speech/Exports";
import {
    IntentRecognitionResult,
    NoMatchReason,
    SpeechRecognitionResult,
    TranslationRecognitionResult,
} from "./Exports";

/**
 * Contains detailed information for NoMatch recognition results.
 * @class NoMatchDetails
 */
export class NoMatchDetails {
    private privReason: NoMatchReason;

    /**
     * Creates and initializes an instance of this class.
     * @constructor
     * @param {NoMatchReason} reason - The no-match reason.
     */
    private constructor(reason: NoMatchReason) {
        this.privReason = reason;
    }

    /**
     * Creates an instance of NoMatchDetails object for the NoMatch SpeechRecognitionResults.
     * @member NoMatchDetails.fromResult
     * @function
     * @public
     * @param {SpeechRecognitionResult | IntentRecognitionResult | TranslationRecognitionResult} result - The recognition result that was not recognized.
     * @returns {NoMatchDetails} The no match details object being created.
     */
    public static fromResult(result: SpeechRecognitionResult | IntentRecognitionResult | TranslationRecognitionResult): NoMatchDetails {
        const simpleSpeech: SimpleSpeechPhrase = SimpleSpeechPhrase.FromJSON(result.json);

        let reason: NoMatchReason = NoMatchReason.NotRecognized;

        switch (simpleSpeech.RecognitionStatus) {
            case RecognitionStatus.BabbleTimeout:
                reason = NoMatchReason.InitialBabbleTimeout;
                break;
            case RecognitionStatus.InitialSilenceTimeout:
                reason = NoMatchReason.InitialSilenceTimeout;
                break;
            default:
                reason = NoMatchReason.NotRecognized;
                break;
        }

        return new NoMatchDetails(reason);
    }

    /**
     * The reason the recognition was canceled.
     * @member NoMatchDetails.prototype.reason
     * @function
     * @public
     * @returns {NoMatchReason} Specifies the reason canceled.
     */
    public get reason(): NoMatchReason {
        return this.privReason;
    }
}
